src/library/clementine/clementineplaylistmodel.cpp

```diff
@@ -0,0 +1,298 @@
+}
+
+QString ClementinePlaylistModel::getFieldString(const QModelIndex& index,
+        const QString& fieldName) const {
+    return getFieldVariant(index, fieldName).toString();
+}
+
+TrackPointer ClementinePlaylistModel::getTrack(const QModelIndex& index) const {
+    QString location = getTrackLocation(index);
+
+    if (location.isEmpty()) {
+        qDebug() << "Track is lost";
+        return TrackPointer();
+    }
+
+    bool trackAlreadyInLibrary = false;
+    TrackPointer pTrack = m_pTrackCollectionManager->getOrAddTrack(
+            TrackRef::fromFileInfo(location),
+            &trackAlreadyInLibrary);
+
+    // If this track was not in the Mixxx library it is now added and will be
+    // saved with the metadata from Clementine. If it was already in the library
+    // then we do not touch it so that we do not over-write the user's metadata.
+    if (pTrack && !trackAlreadyInLibrary) {
+        pTrack->setArtist(getFieldString(index, CLM_ARTIST));
+        pTrack->setTitle(getFieldString(index, CLM_TITLE));
+        pTrack->setDuration(getFieldString(index, CLM_DURATION).toDouble());
+        pTrack->setAlbum(getFieldString(index, CLM_ALBUM));
+        pTrack->setAlbumArtist(getFieldString(index, CLM_ALBUM_ARTIST));
+        pTrack->setYear(getFieldString(index, CLM_YEAR));
+        pTrack->setGenre(getFieldString(index, CLM_GENRE));
+        pTrack->setGrouping(getFieldString(index, CLM_GROUPING));
+        pTrack->setRating(getFieldString(index, CLM_RATING).toInt());
+        pTrack->setTrackNumber(getFieldString(index, CLM_TRACKNUMBER));
+        double bpm = getFieldString(index, CLM_BPM).toDouble();
+        bpm = pTrack->setBpm(bpm);
+        pTrack->setBitrate(getFieldString(index, CLM_BITRATE).toInt());
+        pTrack->setComment(getFieldString(index, CLM_COMMENT));
+        pTrack->setComposer(getFieldString(index, CLM_COMPOSER));
+        // If the track has a BPM, then give it a static beatgrid.
+        if (bpm > 0) {
+            mixxx::BeatsPointer pBeats = BeatFactory::makeBeatGrid(*pTrack, bpm, 0.0);
+            pTrack->setBeats(pBeats);
+        }
+    }
+    return pTrack;
+}
+
+TrackId ClementinePlaylistModel::getTrackId(const QModelIndex& index) const {
+    const auto track = getTrack(index);
+    if (track) {
+        return track->getId();
+    } else {
+        return TrackId();
+    }
+}
+
+// Gets the on-disk location of the track at the given location.
+QString ClementinePlaylistModel::getTrackLocation(const QModelIndex& index) const {
+    if (!index.isValid()) {
+        return QString();
+    }
+    QUrl url(getFieldString(index, CLM_URI));
+
+    QString location;
+    location = url.toLocalFile();
+
+    qDebug() << location << " = " << url;
+
+    if (!location.isEmpty()) {
+        return location;
+    }
+
+    // Try to convert a smb path location = url.toLocalFile();
+    QString temp_location = url.toString();
+
+    if (temp_location.startsWith("smb://")) {
```

Comment:
This is an ugly hack. Do we really need to keep this or better drop SMB support until properly implemented?

Reply:
i removed it as i have no idea about the problem